File: node/agent-transport-livekit/src/observability.ts
/**
 * Observability setup for agent-transport (Node.js).
 *
 * Delegates to livekit-agents' setupCloudTracer() and uploadSessionReport()
 * so we get the same traces, logs, and session reports as standard LiveKit agents.
 *
 * Set LIVEKIT_OBSERVABILITY_URL to enable.
 */

import { createSessionReport, type SessionReportOptions } from '@livekit/agents/dist/voice/report.js';
import { uploadSessionReport } from '@livekit/agents/dist/telemetry/traces.js';
// TODO: Re-enable OTel observability once we finalize the tracing backend.
// import { setupCloudTracer } from '@livekit/agents/dist/telemetry/traces.js';

export function getObservabilityUrl(): string | undefined {
  return process.env.LIVEKIT_OBSERVABILITY_URL;
}

// TODO: Re-enable OTel observability once we finalize the tracing backend.
// let _initialized = false;
//
// export async function setupObservability(callId: string = 'server'): Promise<boolean> {
//   if (_initialized) return true;
//   const obsUrl = getObservabilityUrl();
//   if (!obsUrl) return false;
//   try {
//     await setupCloudTracer({ roomId: callId, jobId: callId, cloudHostname: obsUrl });
//     _initialized = true;
//     console.log(`Observability configured → ${obsUrl}`);
//     return true;
//   } catch {
//     console.warn('Failed to set up cloud tracer — observability disabled.');
//     return false;
//   }
// }

export async function uploadReport(options: {
  agentName: string;
  session: any;
  callId: string;
  recordingPath?: string;
  recordingStartedAt?: number;
}): Promise<void> {
  const obsUrl = getObservabilityUrl();
  if (!obsUrl) return;

  const { agentName, session, callId, recordingPath, recordingStartedAt } = options;

  const reportOpts: SessionReportOptions = {
    jobId: callId,
    roomId: callId,
    room: callId,
    options: session.options ?? {},
    events: session._recordedEvents ?? [],
    chatHistory: session.history?.copy?.() ?? session.history ?? [],
    enableRecording: !!recordingPath,
    startedAt: session._startedAt ?? Date.now(),
    audioRecordingPath: recordingPath,
    audioRecordingStartedAt: recordingStartedAt,
    modelUsage: session.usage?.modelUsage,
  };

  const report = createSessionReport(reportOpts);

  await uploadSessionReport({
    agentName,
    cloudHostname: obsUrl,
    report,
  });
}

File: python/agent_transport/sip/livekit/observability.py
"""Observability setup for agent-transport.

Delegates to livekit-agents' _setup_cloud_tracer() so we get the same
traces, logs, and recording support as standard LiveKit agents.

Set LIVEKIT_OBSERVABILITY_URL to enable.
"""

import logging
import os

logger = logging.getLogger("agent_transport.observability")

_initialized = False


def _get_observability_url() -> str | None:
    return os.environ.get("LIVEKIT_OBSERVABILITY_URL")


def setup_observability(*, call_id: int | str = "server") -> bool:
    """Set up OTel trace and log export via LiveKit's built-in telemetry.

    Returns True if observability was configured, False otherwise.
    """
    global _initialized
    if _initialized:
        return True

    obs_url = _get_observability_url()
    if not obs_url:
        return False

    try:
        from livekit.agents.telemetry.traces import _setup_cloud_tracer
    except ImportError:
        logger.warning(
            "livekit-agents telemetry not available — observability disabled."
        )
        return False

    _setup_cloud_tracer(
        room_id=str(call_id),
        job_id=str(call_id),
        observability_url=obs_url,
    )

    _initialized = True
    logger.info("Observability configured → %s", obs_url)
    return True


def shutdown_observability() -> None:
    """Flush and shut down OTel exporters."""
    global _initialized
    if not _initialized:
        return

    try:
        from livekit.agents.telemetry.traces import _shutdown_telemetry
        _shutdown_telemetry()
    except Exception:
        logger.debug("telemetry shutdown error", exc_info=True)

    _initialized = False
